common/captcha/CaptchaSolverInterface.py

```diff
@@ -0,0 +1,90 @@
+from abc import ABC, abstractmethod
+import os
+import cv2
+import time
+
+
+class SolvedCaptcha:
+    """
+    A class interface for representing solved Captchas.
+    """
+
+    def __init__(self, image, text, answer=None, out_dir=None):
+        """
+        :param image: Unprocessed Captcha image
+        :param text: The text contained in the captcha image. (eg: 12+3=)
+        :param answer: The solved captcha's answer (eg: 15). In the 'jumbled text' captchas this is the same as 'text'.
+        :param out_dir: Where to write captchas to upon notify_correct() or notify_incorrect()
+        """
+        self.image = image
+        self.text = text
+        self.answer = answer
+        self.out_dir = out_dir
+
+    def notify_correct(self):
+        """
+        Writes the Captcha's image to self.correct_dir
+        This should be called in the case a captcha is solved correctly.
+        Captchas should be solved with a unique filename indicating what the captcha says.
+        Eg: YYYYMMDD-HHMMSS_qwerty.png
+        """
+        try:
+            correct_dir = os.path.join(self.out_dir, 'correct')
+            os.makedirs(correct_dir, exist_ok=True)
+            filename = '{}_{}.png'.format(time.strftime("%Y%m%d-%H%M%S"), self.text)
+            cv2.imwrite(os.path.join(correct_dir, filename), self.image)
+        except TypeError:
+            raise NotADirectoryError('notify_incorrect() was called on SolvedCaptcha when no out_dir was given.')
+
+    def notify_incorrect(self):
+        """
+        Writes the Captcha's image to self.incorrect_dir
+        This should be called in the case a captcha is solved incorrectly.
+        Captchas should be solved with a unique filename. Eg: YYYYMMDD-HHMMSS.png
+        """
+        try:
+            incorrect_dir = os.path.join(self.out_dir, 'incorrect')
+            os.makedirs(incorrect_dir, exist_ok=True)
+            filename = '{}.png'.format(time.strftime("%Y%m%d-%H%M%S"))
```

Comment:
Can you include some separators (20200101->2020-01-01, for example)? It's not clear it's a timestamp without them.

Also, you can just do `time.strftime("%Y-%m-%d-%H:%M:%S.png")` here to avoid the double-format.

Reply:
Will do, although colons aren't allowed in filenames ;) Although I think with the YYYY-MM-DD-HHMMSS it's fairly obvious the second part is for the time, so it shouldn't matter.

Reply:
LGTM.